xalia/Win32/HwndButtonProvider.cs

```diff
@@ -178,10 +236,113 @@ public override UiDomValue EvaluateIdentifierLate(UiDomElement element, string i
+                        if (ButtonStateKnown)
+                            return UiDomBoolean.FromBool((ButtonState & BST_INDETERMINATE) != 0);
+                    }
+                    break;
+                case "pushed":
+                    depends_on.Add((element, new IdentifierExpression("win32_button_state")));
+                    if (ButtonStateKnown)
+                        return UiDomBoolean.FromBool((ButtonState & BST_PUSHED) != 0);
+                    break;
+                case "focus":
+                    depends_on.Add((element, new IdentifierExpression("win32_button_state")));
+                    if (ButtonStateKnown)
+                        return UiDomBoolean.FromBool((ButtonState & BST_FOCUS) != 0);
+                    break;
+                case "hot":
+                    depends_on.Add((element, new IdentifierExpression("win32_button_state")));
+                    if (ButtonStateKnown)
+                        return UiDomBoolean.FromBool((ButtonState & BST_HOT) != 0);
+                    break;
+                case "dropdownpushed":
+                    depends_on.Add((element, new IdentifierExpression("win32_button_state")));
+                    if (ButtonStateKnown)
+                        return UiDomBoolean.FromBool((ButtonState & BST_DROPDOWNPUSHED) != 0);
+                    break;
             }
             return UiDomUndefined.Instance;
         }
 
+        public override bool UnwatchProperty(UiDomElement element, GudlExpression expression)
+        {
+            if (expression is IdentifierExpression id)
+            {
+                switch (id.Name)
+                {
+                    case "win32_button_state":
+                        _watchingButtonState = false;
+                        return true;
+                }
+            }
+            return false;
+        }
+
+        public override bool WatchProperty(UiDomElement element, GudlExpression expression)
+        {
+            if (expression is IdentifierExpression id)
+            {
+                switch (id.Name)
+                {
+                    case "win32_button_state":
+                        _watchingButtonState = true;
+                        if (!ButtonStateKnown)
+                            Utils.RunTask(FetchButtonState());
+                        return true;
+                }
+            }
+            return false;
+        }
+
+        private async Task FetchButtonState()
+        {
+            var old_count = _buttonStateChangeCount;
+            int result;
+            try
+            {
+                result = unchecked((int)(long)await SendMessageAsync(Hwnd, BM_GETSTATE, IntPtr.Zero, IntPtr.Zero));
+            }
+            catch (Win32Exception e)
+            {
+                if (!HwndProvider.IsExpectedException(e))
+                    throw;
+                return;
+            }
+            if (_buttonStateChangeCount != old_count)
+                return;

```

Comment:
The scenario this is supposed to account for goes something like this:
 * Property is watched, and fetched successfully.
 * We receive an event saying the property changed, but not saying what it changed to.
 * A query is sent out.
 * The property is unwatched.
 * We receive another event saying the property changed, which does not cause another query to go out because we're no longer watching the property.
 * The query completes, but because it was started before the latest event, it may return a stale value.

At the end of that sequence, the property should be regarded as "unknown", which means that the query needs to detect this situation somehow and not update the value.

Reply:
Hm, yeah that makes sense. I can't really come up with any cleaner way to do this or communicate it, so this is probably fine.